### ISSUE TITLE
Keithley 2000

### ISSUE THREAD

**Issue by AdrianPotter** [ISISComputingGroup/IBEX#2460]
_This ticket will require further refinement (by discussion with the scientist) before it can implemented_

As an instrument scientist, I want to control a Keithley 2000 multimeter, so that I can use it to measure currents and voltages (both AC and DC) as well as resistances (4-wire and 2-wire). 

There may also be additional useful functionality we can expose, we should discuss with electronics & relevant scientists what the use-cases for this device are.

#### Acceptance Criteria
1. I can select the measurement mode:
   1. current, AC or DC
   1. voltage, AC or DC
   1. resistance, 2-wire or 4-wire
1. The system displays the value returned by the multimeter
   1. The system should display the name of the value 
   1. The system should display the value itself
   1. The system should display the units of the value 

#### Notes
1. Discuss detailed requirements with relevant instrument scientists & electronics group before starting this ticket.
1. The Keithley 2000 multimeter can operate in many modes (not just the ones listed above).  How many do we actually need to support?

**Comment by KathrynBaker** [comment]
The command set for this is mainly a subset of the Keithley 2700 (Ticket #647), as such, the two models should be kept in mind when undertaking this ticket. However, it should be noted that #647 may be completed first, which would make this ticket potentially obsolete.

**Comment by kjwoodsISIS** [comment]
Original ticket text:
As an instrument scientist, I would like an IOC et al. for a Kiethley 2000 multimeter so that I can use the one's we've got with EPICS.

This ticket arose because Vicky wanted to use a model 2000 on Iris. We were able to engineer something quickly by taking apart their IOC for the model 2400. The protocols are similar in many ways.

Model 2000s are quite small and apparently are used on instruments sporadically without a great deal of warning. This ticket is not for a specific experiment, but rather to support a common piece of equipment that is often connected.

I haven't looked at the full 2000 spec, but Gareth's intuition is that the IOC should allow for changing the read mode (current/voltage, AC/DC) and then read the subsequent result (via the :READ? command). Different values can be accessed by the :MEAS[...]? command, but this implicitly tells the device to change mode, so we shouldn't have a scan on multiple values.

Talk with Vicky on Iris before embarking on this ticket to see if they need the IOC to do anything else, and let her know once it's done.

**Comment by AdrianPotter** [comment]
The requirement has been met on IRIS using SDTEST. The device is used only sproradically and with limited functionality. I'm proposing we don't create a dedicated IOC for this device at this time given its limited value.

Putting in review to see if that viewpoint is agreed.

**Comment by KathrynBaker** [comment]
It will also be covered on the whole by the 2700, so it may be that we do not need to do anything for this in the end anyway. So I think I can concur that this ticket is of limited value at this time and can be revisited when we know we have a greater range of specific requirements.

**Comment by kjwoodsISIS** [comment]
I was under the impression that the SDTEST solution was a stop-gap.  We've solved the immediate problem on IRIS, but I expect other instruments will want to use Keithley 2000s from time-to-time.  So, having a full solution would be helpful.
If the Keithley 2700 provides that solution, all well and good (see #647).  However, #647 is not explicit about the functionality that will be provided.  It would help to have a statement about what functions the resulting IOC & OPI will perform.

**Comment by AdrianPotter** [comment]
There's an element of that. The understanding I had from Ian though was that they'll only be using the 2000s for a short time, they won't be installed permanently on the beam line. It'd be nice to have a complete catalogue of IOCs for every device, but that's not feasible. The motivation for declaring this a won't fix for the time being is that there's no immediate requirement to provide any additional functionality than what they currently have on IRIS using SDTEST. The ticket requests further refinement of the requirements through discussion with the instrument scientists, but their immediate requirements have been fulfilled.

We could leave this ticket open but, given the lack of immediate need, it is likely to be so low priority as to never be revisited. As you say, other instruments could potentially request to use the device, but that's very speculative. If someone actually needs a 2000, it may be more beneficial to create a new ticket to capture their requirements at the time.

**Comment by AdrianPotter** [comment]
Following discussion with @KathrynBaker, the 2700 solution should also apply to the 2000. Removing won't fix label. Please could we verify this ticket by getting hold of a 2000 and testing that it works with the 2700 IOC. Please could we also ensure there are sufficient 2700s to meet the needs of the instruments. Iris are currently using 4 2000s. Setting to impeded awaiting 2700 ticket to be finished.

**Comment by AdrianPotter** [comment]
The 2700 ticket is #647

**Comment by AdrianPotter** [comment]
Moving back to the backlog. Can be picked up with #647 but isn't in our current plan of work and just sitting idle

**Comment by John-Holt-Tessella** [comment]
in use on ENGINX and has been on use on IRIS let us support this now. SDTEST is not a great solution.

**Comment by LilithCole** [comment]
keithley 2700 is a bit too specialised, a more generic system would be preferable.

**Comment by Tom-Willemsen** [comment]
The use-case on POLREF was for reading a 4-wire resistance - I have added this to acceptance criteria.

**Comment by Tom-Willemsen** [comment]
PRs:
- https://github.com/ISISComputingGroup/EPICS/pull/318
- https://github.com/ISISComputingGroup/ibex_gui/pull/1523
- https://github.com/ISISComputingGroup/EPICS-Keithley_2000/pull/1
- https://github.com/ISISComputingGroup/EPICS-ioc/pull/768
- https://github.com/ISISComputingGroup/IBEX/pull/7588
- Manual added to `\\isis\shares\ISIS_Experiment_Controls\Manuals\Keithley__2000`
- Wiki
